### ISSUE TITLE
cmd/go: impossible to build some packages with `-toolexec` on Windows

### ISSUE THREAD

**Issue by RomainMuller** [golang/go#70046]
### Go version

go version go1.23.2 linux/arm64

### Output of `go env` in your module/workspace:

```shell
GO111MODULE=''
GOARCH='arm64'
GOBIN=''
GOCACHE='/root/.cache/go-build'
GOENV='/root/.config/go/env'
GOEXE=''
GOEXPERIMENT=''
GOFLAGS=''
GOHOSTARCH='arm64'
GOHOSTOS='linux'
GOINSECURE=''
GOMODCACHE='/go/pkg/mod'
GONOPROXY=''
GONOSUMDB=''
GOOS='linux'
GOPATH='/go'
GOPRIVATE=''
GOPROXY='https://proxy.golang.org,direct'
GOROOT='/usr/local/go'
GOSUMDB='sum.golang.org'
GOTMPDIR=''
GOTOOLCHAIN='local'
GOTOOLDIR='/usr/local/go/pkg/tool/linux_arm64'
GOVCS=''
GOVERSION='go1.23.2'
GODEBUG=''
GOTELEMETRY='local'
GOTELEMETRYDIR='/root/.config/go/telemetry'
GCCGO='gccgo'
GOARM64='v8.0'
AR='ar'
CC='gcc'
CXX='g++'
CGO_ENABLED='1'
GOMOD='/Users/romain.marcadier/Development/Datadog/orchestrion/go.mod'
GOWORK=''
CGO_CFLAGS='-O2 -g'
CGO_CPPFLAGS=''
CGO_CXXFLAGS='-O2 -g'
CGO_FFLAGS='-O2 -g'
CGO_LDFLAGS='-O2 -g'
PKG_CONFIG='pkg-config'
GOGCCFLAGS='-fPIC -pthread -Wl,--no-gc-sections -fmessage-length=0 -ffile-prefix-map=/tmp/go-build89119916=/tmp/go-build -gno-record-gcc-switches'
```

### What did you do?

The following command works as intended:

```console
$ go build -a github.com/elastic/go-elasticsearch/v8/typedapi/types
```

However if you use any `-toolexec` value, it suddenly fails on Windows:
```console
$ go build -toolexec="D:\\a\\orchestrion\\orchestrion\\bin\\orchestrion.exe" -a github.com/elastic/go-elasticsearch/v8/typedapi/types
fork/exec D:\\a\\orchestrion\\orchestrion\\bin\\orchestrion.exe: The filename or extension is too long.
```

### What did you see happen?

What we observe is that when there is no `-toolexec` the Windows arguments size limit (of 32KiB) is "worked around" by passing an `@args-file` value that contains all the arguments instead of passing the arguments list itself directly...

Unfortunately, the use of `-toolexec` appears to disable this mechanism, making it impossible to build packages with very large amounts of files in them on Windows.

### What did you expect to see?

I expect a package that successfully builds without `-toolexec` to also be build-able with `-toolexec`, and for the `-toolexec` tool to actually also receive the arguments file stand-in when appropriate.

**Comment by RomainMuller** [comment]
As suggested by @gabyhelp, this is actually very much related to (although this has the added characteristic of being an apparent `-toolexec` break out of the response file feature):
- https://github.com/golang/go/issues/18468
- https://github.com/golang/go/issues/28372
- https://github.com/golang/go/issues/60183
- https://github.com/golang/go/issues/38569
- https://github.com/golang/go/issues/44782

**Comment by RomainMuller** [comment]
Also, the culprit is likely this code path:

https://github.com/golang/go/blob/e5e552b816793f2b5729744c27995cce12baf52e/src/cmd/go/internal/work/exec.go#L3431-L3435

Since in our case, the `prog` ends up being the `-toolexec` executable...

**Comment by prattmic** [comment]
cc @matloob @samthanawalla

**Comment by ianlancetaylor** [comment]
It's fair to say that most Unix programs do not understand the `@` syntax for response files.  So it wouldn't be appropriate to use it by default.

Do most Windows programs understand that syntax?

**Comment by RomainMuller** [comment]
I agree - this could be surprising to a bunch of tools; although... On Windows specifically, if the CLI arguments exceed 32KiB, then the command will basically just not run... It'll just fail differently if you pass a response file instead of the arguments list in that case...

But regardless; I think it would be a-okay (to me at least) to either have:
- Another CLI flag to signal the `-toolexec` command understands response files
- A "special" syntax in `-toolexec` that signifies it understands response files

It might also be relevant to document the syntax of these files publicly so that it feels less like copying CLI implementation details (apparently it supports line continuations, the response file can contain further response files, etc...)

**Comment by RomainMuller** [comment]
I will add that today, if the `-toolexec` tool is named `compile`, `link`, `cgo`, `asm`, or `cover`, it'll receive the response file (and possibly crash if it does not support this). I have been able to verify this with `orchestrion` today.

---

In addition to this, I would say that today, any change to any tool's arguments could be seen as a breaking change for `-toolexec` tools that process said arguments... I think the contract of `-toolexec` is that it either must not care about the arguments being passed (just forwarding them as-is), or it must be ready to deal with _anything_ the tool may receive; including a response file.

From this perspective, whether sending the `@response-file` arguments via `-toolexec` tools is a breaking change or not is somewhat debatable.

**Comment by mvdan** [comment]
> I think the contract of `-toolexec` is that it either must not care about the arguments being passed (just forwarding them as-is), or it must be ready to deal with _anything_ the tool may receive; including a response file.

That is what I have been assuming with my own development of toolexec tools.

FWIW I already wrote generic response file support in Go at https://pkg.go.dev/mvdan.cc/responsefile, and you're welcome to reuse it as well.

**Comment by matloob** [comment]
> > I think the contract of `-toolexec` is that it either must not care about the arguments being passed (just forwarding them as-is), or it must be ready to deal with _anything_ the tool may receive; including a response file.
> 
> That is what I have been assuming with my own development of toolexec tools.

I would have made the same assumption.

I'm wondering if there are any tools that already exist that examine the arguments passed in used in workflows that would break if we decided to start passing in response file arguments?

**Comment by mvdan** [comment]
I don't think any toolexec tools written today would handle @-arg response files, because they simply have never been supplied before. But those same tools are already pretty broken today when very many files are being built at once, e.g. https://github.com/burrowers/garble/issues/572, so I think we need to cross this bridge sooner or later.

**Comment by matloob** [comment]
Ah. This seems reasonable to me.

**Comment by mvdan** [comment]
@matloob I personally don't think a proposal is needed here. `-toolexec` is not a documented and stable protocol of any kind, and if anything, this just fixes a bug and improves consistency.
